### PR TITLE
Minor java improvements

### PR DESCRIPTION
**Pull request by djh82**
## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `CHANGES.txt` (and read the `README.rst`)
* [x] I have updated the appropriate documentation

Here are some minor Java improvements for discussion; I've yet to add tests and update the CHANGES.txt/relevant documentation, but would be happy to do so if there it is decided that these changes are worth integrating.

### CODE REVIEW

**Comment by bdbaddog** [comment]
Before I dig around the code, please summarize the improvements (with an eye towards being able to review the PR)?
(rather than me having to figure out what the change is supposed to do?)

**Comment by djh82** [comment]
This PR aggregates 3 of the changes we've made to improve java handling:

1. Introduces a scanner which handles both JAVACLASSPATH ~~and JAVABOOTCLASSPATH~~, otherwise we have to litter our code with Depends. 
2. Improves classpath generation, since if you added a nodelist to the JAVA*PATH env variable, you'd end up with `a.jar:[b.jar]`
3. ~~Tries to improve the handling of generated source files to prevent constant rebuilding. Currently scons assumes generated files will create class files in the default package, but 9 times out of 10 this is unlikely to be the case. The standard (but not required) is to put generated source files on their package directory, and hence this change tries to deduce that. Java protobuf is a good example. Note that this could still get it wrong, but it's no less wrong, and it's probably right more of the time.~~

**Comment by djh82** [comment]
#3761 has a partial fix for this

**Comment by bdbaddog** [comment]
@davidhart82 - can you add some tests to cover these changes?

**Comment by djh82** [comment]
@bdbaddog I've scrubbed fix number 3; it might need more work.  But I've added tests for the other 2 fixes.

**Comment by mwichmann** [comment]
Is there a replacement coming?

File: SCons/Scanner/Java.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+def scan(node, env, libpath=()):
+    classpath = env.get('JAVACLASSPATH', [])
+    classpath = _subst_libs(env, classpath)
+
+    result = []
+    for lib in classpath:
+        if SCons.Util.is_String(lib) and "*" in lib:
+            result += env.Glob(lib)
+        elif os.path.isdir(str(lib)):
```
Comment (djh82) [inline SCons/Scanner/Java.py]: I need to find all of the in-memory and on-disk nodes; is there a better way of doing this?

**Comment by djh82** [comment]
> Is there a replacement coming?

@mwichmann I got fed up with it, and was going to bin it.  But I had a change of heart, reduced scope a little, and now I think it works.

File: SCons/Scanner/Java.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    The classpath can contain:
+     - Explicit paths to JAR/Zip files
+     - Wildcards (*)
+     - Directories which contain classes in an unnamed package
+     - Parent directories of the root package for classes in a named package
+
+     Class path entries that are neither directories nor archives (.zip or JAR files) nor the asterisk (*) wildcard character are ignored.
+     """
+    classpath = env.get('JAVACLASSPATH', [])
```
Comment (djh82) [inline SCons/Scanner/Java.py:68]: This might work for JAVABOOTCLASSPATH now

File: SCons/Scanner/JavaTests.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        deps_match(self, deps, expected)
+
+
+class JavaScannerDirClasspath(unittest.TestCase):
+    def runTest(self):
+        env = DummyEnvironment(JAVASUFFIXES=['.java'],
+                               JAVACLASSPATH=[test.workpath()])
+        s = SCons.Scanner.Java.JavaScanner()
+        deps = s(DummyNode('dummy'), env)
+        expected = ['Test.class', 'com/Test.class']
```
Comment (djh82) [inline SCons/Scanner/JavaTests.py:148]: Check if this should pick up jars too

File: SCons/Scanner/Java.py (hunk 1 above)
Comment (bdbaddog) [reply SCons/Scanner/Java.py]: SCons Glob should find both..

Comment (djh82) [reply SCons/Scanner/Java.py]: The age old scons glob isn't recursive problem - need to find classes in the current directory and below.

**Comment by mwichmann** [comment]
This one has set open for a while now - what do we need to do to get it over the line? I'd probably change the title a little bit - adding a scanner feels like a little more than just "minor improvements".  I pulled the PR to here and rebased it to head and the added tests still pass fine. I'm no Java person so not much further help from me.

A few nitpickery notes for updates:
1. Run Black on the two added files - new files might as well enter the picture in the expected eventual style (I know it doesn't say this anywhere!)
2. The CHANGES.txt snip needs to move to the next-release location, a rebase leaves it where it was (under 4.2.0)
3. Seems like a new scanner should get a snip in RELEASE.txt
4. Check for trailing whitespace

**Comment by djh82** [comment]
@mwichmann I'd almost given up on it! We use this code everyday, so am happy to fix it up inline with your suggestions. Hopefully I'll get some time this week to do that.

**Pull request merged**